### DISCUSSION TITLE
Allow file upload

### DISCUSSION THREAD

**Discussion by shiipou** [lenra-io/lenra_cli#55] [General]
# Files upload functionality

To prepare the Lenra beta release, we need to have the possibility for user to upload files to app they use.

We must handle 3 part for this functionality : 
- The client-side to allow user to select a file and send it to the server
- The server-side to allow dev to get the file data
- The storage side to save the file

## Client-side

For the client-side, I think we can use the file browser that work for many platform : iOS, Linux, macOS, web, Windows
For android we need to provide an alternative or write a simple package. (ex: [path_selector](https://pub.dev/packages/path_selector))

## Server-side

We can send the data to the listener and let the listener save the file content to the storage.
Or we can save the file to the storage server and provide the "path" to the listener to allow listener to read it.

We also must provide some API endpoint to allow the app to write/read/delete the file on the storage.

## Storage-side

We can bind a k8s volume to the app container to allow the app to access data, but that can have some limitation (blocking scalability, replica can only be run on one k8s node if they share the same volume)

We can also host an S3 cluster (Or rent one) to save files using an HTTP API. This solution seems better for me, but we need to check if there is no app's use case blocked by this solution.

**Comment by Nesqwik** [comment]
For the storage, IMO the most simple way to do this is to use the [OVH Object storage](https://www.ovhcloud.com/fr/public-cloud/object-storage/).
For the server part, both are possible solutions. The first one allow the developper to update the file (compress/resize the image for example) before store it and setup some size limitations. 
Otherwise, we should add this feature to the API that store the file (IMO we usually cannot store directly the images/video without transformations/compression first.)
